File: AlgorithmsCpp/数据结构/836_合并集合.cpp
# include<iostream>

using namespace std;

const int N = 100010;
int n, m;
int q[N];

int find(int x) {
	if (q[x] != x) q[x] = find(q[x]);
	return q[x];
}

int main()
{
	cin >> n >> m;
	for (int i = 0; i < n; i++) {
		q[i] = i;
	}

	while (m--) {
		char op[2];
		int a, b;
		scanf("%s%d%d", op, &a, &b);
		if (*op == 'M') {
			q[find(a)] = find(b);
		}
		else {
			if (find(a) == find(b)) {
				printf("Yes\n");
			}
			else {
				printf("No\n");
			}
		}
	}

	return 0;
}

File: AlgorithmsCSharp/数据结构/836_合并集合.cs
namespace _836_合并集合
{
    class Program
    {
        static int[] p;

        static void Main()
        {
            int n = Convert.ToInt32(Console.ReadLine());
            int m = Convert.ToInt32(Console.ReadLine());
            p = new int[100010];
            for (int i = 0; i < n; i++)
            {
                p[i] = i;
            }

            while (m-- > 0)
            {
                string op = Console.ReadLine();
                int a = Convert.ToInt32(Console.ReadLine());
                int b = Convert.ToInt32(Console.ReadLine());
                if (op == "M")
                {
                    p[Find(a)] = Find(b);
                }
                else
                {
                    if (Find(a) == Find(b)) Console.WriteLine("Yes");
                    else Console.WriteLine("No");
                }
            }
        }

        static int Find(int x)
        {
            if (p[x] != x) p[x] = Find(p[x]);
            return p[x];
        }
    }
}
